File: src/main/java/actions/KeyHandler.java
package actions;

import processing.core.PApplet;
import processing.event.KeyEvent;
import ux.DrawRateManager;

import java.util.*;
import java.util.stream.Collectors;

public class KeyHandler {
    private final Map<Set<KeyCombo>, KeyCallback> keyCallbacks;
    private final static Set<Integer> pressedKeys = new HashSet<>();

    public static Set<Integer> getPressedKeys() {
        return Collections.unmodifiableSet(pressedKeys);
    }

    private KeyHandler(Builder builder) {
        this.keyCallbacks = builder.keyCallbacks;
        builder.processing.registerMethod("keyEvent", this);
    }

    private Set<KeyCombo> findDuplicateKeyCombos(Set<KeyCombo> keyCombos) {
        Set<KeyCombo> duplicateKeyCombos = new HashSet<>();

        for (Map.Entry<Set<KeyCombo>, KeyCallback> entry : keyCallbacks.entrySet()) {
            Set<KeyCombo> existingKeyCombos = entry.getKey();
            for (KeyCombo keyCombo : keyCombos) {
                if (existingKeyCombos.contains(keyCombo)) {
                    duplicateKeyCombos.add(keyCombo);
                }
            }
        }

        return duplicateKeyCombos;
    }

    public void keyEvent(KeyEvent event) {
        int keyCode = event.getKeyCode();

        for (KeyCallback callback : keyCallbacks.values()) {
            if (callback.matches(event)) {

                if (event.getAction() == KeyEvent.PRESS) {
                    pressedKeys.add(keyCode);
                    callback.invokeFeature();
                    callback.notifyKeyObservers();
                    ;
                    DrawRateManager.getInstance().drawImmediately();
                }

                if (event.getAction() == KeyEvent.RELEASE) {
                    pressedKeys.remove(keyCode);
                    callback.cleanupFeature();
                }

                break;
            }
        }
    }

    public String getUsageText() {
        StringBuilder usageText = new StringBuilder("Key Usage:\n\n");

        Set<KeyCallback> processedCallbacks = new HashSet<>();

        int maxKeysWidth = keyCallbacks.values().stream()
                .filter(KeyCallback::isValidForCurrentOS)
                .mapToInt(callback -> callback.getValidKeyCombosForCurrentOS().stream()
                        .map(KeyCombo::toString)
                        .collect(Collectors.joining(", ")).length())
                .max().orElse(0);


        for (KeyCallback callback : keyCallbacks.values()) {
            // valid for OS and not already processed
            if (!callback.isValidForCurrentOS() || processedCallbacks.contains(callback)) {
                continue;
            }

            // Only include key combos that are valid for the current OS
            String keysString = callback.getValidKeyCombosForCurrentOS().stream()
                    .map(KeyCombo::toString)
                    .collect(Collectors.joining(", "));

            String usageDescription = callback.getUsageText();
            usageText.append(String.format("%-" + (maxKeysWidth + 1) + "s: %s\n", keysString, usageDescription));

            processedCallbacks.add(callback);
        }

        return usageText.toString();
    }

    // Here's our Builder class
    public static class Builder {
        private final Map<Set<KeyCombo>, KeyCallback> keyCallbacks = new LinkedHashMap<>();
        private final PApplet processing;

        public Builder(PApplet processing) {
            this.processing = processing;
        }

        public Builder addKeyCallback(KeyCallback callback) {
            Set<KeyCombo> keyCombos = callback.getKeyCombos();
            Set<KeyCombo> duplicateKeyCombos = findDuplicateKeyCombos(keyCombos);

            if (!duplicateKeyCombos.isEmpty()) {
                String duplicateKeyCombosString = duplicateKeyCombos.stream()
                        .map(KeyCombo::toString)
                        .collect(Collectors.joining(", "));
                throw new IllegalArgumentException("The following key combos are already associated with another callback: " + duplicateKeyCombosString);
            }

            keyCallbacks.put(keyCombos, callback);
            return this; // this allows us to chain method calls
        }

        private Set<KeyCombo> findDuplicateKeyCombos(Set<KeyCombo> keyCombos) {
            Set<KeyCombo> duplicateKeyCombos = new HashSet<>();

            for (Map.Entry<Set<KeyCombo>, KeyCallback> entry : keyCallbacks.entrySet()) {
                Set<KeyCombo> existingKeyCombos = entry.getKey();
                for (KeyCombo keyCombo : keyCombos) {
                    if (existingKeyCombos.contains(keyCombo)) {
                        duplicateKeyCombos.add(keyCombo);
                    }
                }
            }

            return duplicateKeyCombos;
        }

        public KeyHandler build() {
            return new KeyHandler(this);
        }
    }
}

File: src/main/kotlin/actions/KeyHandler.kt
package actions

import processing.core.PApplet
import processing.event.KeyEvent
import ux.DrawRateManager

class KeyHandler private constructor(builder: Builder) {
    private val keyCallbacks: Map<Set<KeyCombo>, KeyCallback>

    init {
        keyCallbacks = builder.keyCallbacks
        // this registration is what keeps a reference of KeyHandler around
        // for the duration of the program even though you don't explicitly
        // store it anywhere else...
        builder.processing.registerMethod("keyEvent", this)
    }

    @Suppress("unused")
    fun keyEvent(event: KeyEvent) {
        val keyCode = event.keyCode
        // elvis operator - if not null, assign to matchingCallback - if null, return
        val matchingCallback = keyCallbacks.values.find { it.matches(event) } ?: return

        if (event.action == KeyEvent.PRESS) {
            _pressedKeys.add(keyCode)
            matchingCallback.invokeFeature()
            matchingCallback.notifyKeyObservers()
            DrawRateManager.instance!!.drawImmediately()
        }
        if (event.action == KeyEvent.RELEASE) {
            _pressedKeys.remove(keyCode)
            matchingCallback.cleanupFeature()
        }
    }

    val usageText: String
        get() {
            val usageText = StringBuilder("\nKey Usage:\n")
            val processedCallbacks: MutableSet<KeyCallback> = mutableSetOf()

            val maxKeysWidth = keyCallbacks.values.filter { it.isValidForCurrentOS }.maxOfOrNull { callback ->
                callback.validKeyCombosForCurrentOS.joinToString(", ").length
            } ?: 0

            keyCallbacks.values.filter { callback ->
                callback.isValidForCurrentOS && !processedCallbacks.contains(callback)
            }.forEach { callback ->
                val keysString = callback.validKeyCombosForCurrentOS.joinToString(", ")
                val usageDescription = callback.getUsageText()
                usageText.append("${keysString.padEnd(maxKeysWidth + 1)}: $usageDescription\n")
                processedCallbacks.add(callback)
            }

            return usageText.toString()
        }

    class Builder(val processing: PApplet) {
        val keyCallbacks: MutableMap<Set<KeyCombo>, KeyCallback> = mutableMapOf()

        fun addKeyCallback(callback: KeyCallback): Builder {
            val keyCombos = callback.keyCombos
            val duplicateKeyCombos = findDuplicateKeyCombos(keyCombos)
            require(duplicateKeyCombos.isEmpty()) {
                "The following key combos are already associated with another callback: ${duplicateKeyCombos.joinToString(", ")}"
            }
            keyCallbacks[keyCombos] = callback
            return this
        }

        private fun findDuplicateKeyCombos(keyCombos: Set<KeyCombo>): Set<KeyCombo> {
            return keyCallbacks.keys.flatten().intersect(keyCombos)
        }

        fun build(): KeyHandler = KeyHandler(this)
    }

    companion object {
        private val _pressedKeys: MutableSet<Int> = mutableSetOf()
        val pressedKeys: Set<Int> get() = _pressedKeys
    }
}